pkg/reconciler/apis/apiexport/apiexport_reconcile.go

```diff
@@ -0,0 +1,175 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiexport
+
+import (
+	"context"
+	cryptorand "crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/third_party/conditions/util/conditions"
+)
+
+func (c *controller) reconcile(ctx context.Context, apiExport *apisv1alpha1.APIExport) error {
+	identity := apiExport.Spec.Identity
+	if identity == nil {
+		identity = &apisv1alpha1.Identity{}
+	}
+
+	clusterName := logicalcluster.From(apiExport)
+
+	if identity.SecretRef == nil {
+		c.ensureSecretNamespaceExists(ctx, clusterName)
+
+		// See if the generated secret already exists (for whatever reason)
+		_, err := c.getSecret(ctx, clusterName, c.secretNamespace, apiExport.Name)
+		if err != nil && !errors.IsNotFound(err) {
+			return fmt.Errorf("error checking if APIExport %s|%s identity secret %s|%s/%s exists: %w",
+				clusterName, apiExport.Name,
+				clusterName, c.secretNamespace, apiExport.Name,
+				err,
+			)
+		}
+		if errors.IsNotFound(err) {
+			if err := c.createIdentitySecret(ctx, clusterName, apiExport.Name); err != nil {
+				conditions.MarkFalse(
+					apiExport,
+					apisv1alpha1.APIExportIdentityValid,
+					apisv1alpha1.IdentityGenerationFailedReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"Error creating identity secret: %v",
+					err,
+				)
+
+				return err
+			}
+		}
+
+		identity.SecretRef = &corev1.SecretReference{
+			Namespace: c.secretNamespace,
+			Name:      apiExport.Name,
+		}
+
+		apiExport.Spec.Identity = identity
+
+		// Record the spec change. A future iteration will store the hash in status.
+		return nil
+	}
+
+	// Ref exists - make sure it's valid
+	if err := c.updateOrVerifyIdentitySecretHash(ctx, clusterName, apiExport); err != nil {
+		conditions.MarkFalse(
+			apiExport,
+			apisv1alpha1.APIExportIdentityValid,
+			apisv1alpha1.IdentityVerificationFailedReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+	}
+
+	return nil
+}
+
+func (c *controller) ensureSecretNamespaceExists(ctx context.Context, clusterName logicalcluster.LogicalCluster) {
+	if _, err := c.getNamespace(clusterName, c.secretNamespace); errors.IsNotFound(err) {
```

Comment:
no action needed: a direct create would do too.

Reply:
When I was working on Velero, we had a user complain because GCP/GKE alerts were triggering every time Velero tried to create its namespace (the namespace had been precreated and the Velero SA didn't have permission to create it). So I tend to do get-then-create now 😄